[aotd] Support mutations in reordering_to_mimic_autograd_engine

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #155354
* __->__ #155353

Original issue: https://github.com/pytorch/pytorch/issues/154820

Dedicated sub-issue: https://github.com/pytorch/pytorch/issues/155242

Backward graph is reordered by partitioners.py: reordering_to_mimic_autograd_engine

Which only records in the backward graph compute that starts from tangents.

Mutation of primals(inputs) in backward can be disconnected from backward.

Handling this copy_ specifically, as we  add this mutation in framework and this is the only mutation that exist.

IvanKobzarev committed:
[aotd] Support mutations in reordering_to_mimic_autograd_engine

[ghstack-poisoned]

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -7779,6 +7779,38 @@ def run(f):
+        torch.compile(fn, backend="aot_eager", fullgraph=True)(
+            dummy, inplace
+        ).sum().backward()
+        self.assertEqual(ref, inplace)
```
can we have the test assert that the inputs are correct, both:

(1) after running the compiled (and reference) forwards, but before running the backward

(2) after running the backward?

That should help ensure that the test actually confirms that we are not e.g. moving the backward mutation into the forward graph, too

bdhirsh reviewed (approved):
lgtm minus test nit!

IvanKobzarev committed:
Update on "[aotd] Support mutations in reordering_to_mimic_autograd_engine"


Original issue: https://github.com/pytorch/pytorch/issues/154820

Dedicated sub-issue: https://github.com/pytorch/pytorch/issues/155242

Backward graph is reordered by partitioners.py: reordering_to_mimic_autograd_engine

Which only records in the backward graph compute that starts from tangents.

Mutation of primals(inputs) in backward can be disconnected from backward.

Handling this copy_ specifically, as we  add this mutation in framework and this is the only mutation that exist.




[ghstack-poisoned]
